pgjdbc/src/main/java/org/postgresql/largeobject/BlobOutputStream.java

```diff
@@ -48,40 +51,67 @@ public BlobOutputStream(LargeObject lo) {
    * Create an OutputStream to a large object.
    *
    * @param lo LargeObject
-   * @param bsize The size of the buffer used to improve performance
+   * @param bufferSize The size of the buffer for single-byte writes
    */
-  public BlobOutputStream(LargeObject lo, int bsize) {
+  public BlobOutputStream(LargeObject lo, int bufferSize) {
     this.lo = lo;
-    this.bsize = bsize;
-    buf = new byte[bsize];
-    bpos = 0;
+    this.bufferSize = bufferSize;
   }
 
   public void write(int b) throws java.io.IOException {
     LargeObject lo = checkClosed();
-    try {
-      if (bpos >= bsize) {
+    try (ResourceLock ignore = lock.obtain()) {
+      byte[] buf = this.buf;
+      if (buf == null) {
+        // Allocate the buffer on the first use
+        this.buf = buf = new byte[bufferSize];
+      }
+      buf[bufferPosition++] = (byte) b;
+      if (bufferPosition >= bufferSize) {
         lo.write(buf);
-        bpos = 0;
+        bufferPosition = 0;
       }
-      buf[bpos++] = (byte) b;
     } catch (SQLException se) {
       throw new IOException(se.toString());
     }
   }
 
-  public void write(byte[] buf, int off, int len) throws java.io.IOException {
-    LargeObject lo = checkClosed();
-    try {
-      // If we have any internally buffered data, send it first
-      if (bpos > 0) {
-        flush();
+  public void write(byte[] b, int off, int len) throws java.io.IOException {
+    try (ResourceLock ignore = lock.obtain()) {
+      LargeObject lo = checkClosed();
+      byte[] buf = this.buf;
+      // Initialize the buffer if needed
+      // Suppose we have bufferPosition=10KiB buffered out of bufferSize=64KiB, and the user issues
+      // 100KiB write. We will copy 54KiB to the buffer and flush it, then we will copy
+      // the remaining 46KiB to the buffer without flushing to the DB. It avoids small writes
+      // hitting the database.
+      // If the incoming request is 300KiB, then we copy 54KiB to the buffer and flush it.
+      // Then we write the remaining 246KiB directly to the database.
+      // At worst, it results in two DB calls
+      // If the buffer was not there, we do not create one if the write request is big enough
+      if (buf == null && len < 2 * bufferSize) {
+        this.buf = buf = new byte[bufferSize];
       }
```

Comment:
This does not play well for small blobs. For instance, if the user creates 100 byte-long blob, it would be a waste to create 128KiB buffer and discard it right after flush/close. We probably need a dynamically increasing buffer size here.

Reply:
Not sure how we would ever know enough to do this. One thought is to just look at the size of the incoming b[] as a hint as to how big to make the buffer?

Reply:
That is right. I treat "buffer size" as the "max buffer size", and I use "min(incomingSize * 2, maxBufferSize)" as a sizing hint.